File: browser-switch/src/main/java/com/braintreepayments/api/ChromeCustomTabsInternalClient.java
package com.braintreepayments.api;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.widget.Toast;

import androidx.activity.ComponentActivity;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.OptIn;
import androidx.annotation.VisibleForTesting;
import androidx.browser.auth.AuthTabIntent;
import androidx.browser.customtabs.CustomTabsIntent;
import static androidx.browser.customtabs.CustomTabsIntent.OPEN_IN_BROWSER_STATE_OFF;
import androidx.browser.customtabs.ExperimentalOpenInBrowser;
import androidx.browser.trusted.TrustedWebActivityIntentBuilder;
import androidx.core.app.ActivityOptionsCompat;

class ChromeCustomTabsInternalClient {

    private final CustomTabsIntent.Builder customTabsIntentBuilder;
    private final AuthTabIntent.Builder authTabIntentBuilder;
    private ActivityResultLauncher<Intent> launcher;

    ChromeCustomTabsInternalClient(@NonNull ComponentActivity activity) {
        this(new CustomTabsIntent.Builder());
        launcher = AuthTabIntent.registerActivityResultLauncher(activity, this::handleAuthResult);
    }

    @VisibleForTesting
    ChromeCustomTabsInternalClient(CustomTabsIntent.Builder builder) {
        this.customTabsIntentBuilder = builder;
        this.authTabIntentBuilder = new AuthTabIntent.Builder();
    }

    void handleAuthResult(AuthTabIntent.AuthResult result) {
        String message = null;
        switch (result.resultCode) {
            case AuthTabIntent.RESULT_OK: message = "Received auth result."; break;
            case AuthTabIntent.RESULT_CANCELED: message = "AuthTab canceled."; break;
            case AuthTabIntent.RESULT_VERIFICATION_FAILED: message = "Verification failed."; break;
            case AuthTabIntent.RESULT_VERIFICATION_TIMED_OUT: message = "Verification timed out."; break;
            case AuthTabIntent.RESULT_UNKNOWN_CODE: message = "Unknown result code."; break;
        }

        if (result.resultCode == AuthTabIntent.RESULT_OK) {
            message += " Uri: " + result.resultUri;
        }

//        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        Log.d("AuthTab = message is ", message);
    }

    @OptIn(markerClass = ExperimentalOpenInBrowser.class)
    void launchUrl(@NonNull ComponentActivity activity, Uri url, LaunchType launchType) throws ActivityNotFoundException {
//        ActivityResultLauncher<Intent> launcher = AuthTabIntent.registerActivityResultLauncher(activity, this::handleAuthResult);
        AuthTabIntent customTabsIntent = authTabIntentBuilder
                .setEphemeralBrowsingEnabled(true)
//                .setOpenInBrowserButtonState(OPEN_IN_BROWSER_STATE_OFF) // this was requested by a merchant, should
//                .setCloseButtonEnabled(false)
//                .setBackgroundInteractionEnabled(false)
//                .setCloseButtonPosition(CustomTabsIntent.CLOSE_BUTTON_POSITION_END)
//                .setToolbarCornerRadiusDp(10)
                // probably be a setting exposed to customize by the merchant.
                .build();
        if (launchType != null) {
            switch (launchType) {
                case ACTIVITY_NEW_TASK:
                    customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    break;
                case ACTIVITY_CLEAR_TOP:
                    customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    break;
            }
        }
        customTabsIntent.launch(launcher, url, url.getPathSegments().get(0));
    }


//    @OptIn(markerClass = ExperimentalOpenInBrowser.class)
//    void launchUrl(Context context, Uri url, LaunchType launchType) throws ActivityNotFoundException {
//        CustomTabsIntent customTabsIntent = customTabsIntentBuilder
//                .setEphemeralBrowsingEnabled(true)
//                .setOpenInBrowserButtonState(OPEN_IN_BROWSER_STATE_OFF) // this was requested by a merchant, should
//                .setCloseButtonEnabled(false)
//                .setBackgroundInteractionEnabled(false)
//                .setCloseButtonPosition(CustomTabsIntent.CLOSE_BUTTON_POSITION_END)
//                .setToolbarCornerRadiusDp(10)
//                // probably be a setting exposed to customize by the merchant.
//                .build();
//        if (launchType != null) {
//            switch (launchType) {
//                case ACTIVITY_NEW_TASK:
//                    customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
//                    break;
//                case ACTIVITY_CLEAR_TOP:
//                    customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
//                    break;
//            }
//        }
//        customTabsIntent.launchUrl(context, url);
//    }
}

File: browser-switch/src/main/java/com/braintreepayments/api/ChromeCustomTabsInternalClient.kt
package com.braintreepayments.api

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.ActivityNotFoundException
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.RemoteException
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.annotation.OptIn
import androidx.browser.customtabs.CustomTabsCallback
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsServiceConnection
import androidx.browser.customtabs.EngagementSignalsCallback
import androidx.browser.customtabs.ExperimentalMinimizationCallback


internal class ChromeCustomTabsInternalClient(
    private val customTabsIntentBuilder: CustomTabsIntent.Builder = CustomTabsIntent.Builder()
) {
    private var customTabsServiceConnection: CustomTabsServiceConnection? = null
    private var targetActivity: ComponentActivity? = null

    @Throws(ActivityNotFoundException::class)
    fun launchUrl(
        context: Context,
        url: Uri,
        launchAsNewTask: Boolean
    ) {
        val customTabsIntent = customTabsIntentBuilder.build()
        if (launchAsNewTask) {
            customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        customTabsIntent.launchUrl(context, url)
    }

    @OptIn(ExperimentalMinimizationCallback::class)
    fun openCustomTab(
        activity: ComponentActivity,
        url: Uri,
        launchAsNewTask: Boolean,
        browserSwitchCallbacks: BrowserSwitchCallbacks?
    ) {
        targetActivity = activity
        val customTabsCallback = object : CustomTabsCallback() {
            override fun onNavigationEvent(navigationEvent: Int, extras: Bundle?) {
                super.onNavigationEvent(navigationEvent, extras)

                when (navigationEvent) {
                    TAB_SHOWN -> Log.d("asdf", "tab shown")
                    TAB_HIDDEN -> Log.e("asdf", "TAB_HIDDEN")
                }
            }

            override fun onMinimized(extras: Bundle) {
                super.onMinimized(extras)
                browserSwitchCallbacks?.onMinimized?.invoke()
            }
        }

        val customTabsServiceConnection = object : CustomTabsServiceConnection() {
            override fun onServiceDisconnected(name: ComponentName?) {
                Log.d("asdf", "Service Disconnected")
            }

            @SuppressLint("RequiresFeature")
            override fun onCustomTabsServiceConnected(name: ComponentName, client: CustomTabsClient) {
                val customTabsSession = client.newSession(customTabsCallback)
                client.warmup(0L)

                try {
                    val engagementSignalsApiAvailable = customTabsSession?.isEngagementSignalsApiAvailable(Bundle.EMPTY)
                    if (engagementSignalsApiAvailable != true) {
                        Log.d(
                            "asdf", "CustomTab Engagement signals not available, make sure to use the " +
                                    "latest Chrome version and enable via chrome://flags/#cct-real-time-engagement-signals"
                        )
                        return
                    }
                    customTabsSession.setEngagementSignalsCallback(
                        getEngagementSignalsCallback(browserSwitchCallbacks),
                        Bundle.EMPTY
                    )
                } catch (e: RemoteException) {
                    Log.w("asdf", "The Service died while responding to the request.", e)
                } catch (e: UnsupportedOperationException) {
                    Log.w("asdf", "Engagement Signals API isn't supported by the browser.", e)
                }

                val customTabsIntent = CustomTabsIntent.Builder(customTabsSession).build()
                if (launchAsNewTask) {
                    customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
                customTabsIntent.launchUrl(activity, url)
            }
        }

        val packageName = "com.android.chrome"
        val isBound = CustomTabsClient.bindCustomTabsService(activity, packageName, customTabsServiceConnection)
        if (!isBound) {
            Log.d("asdf", "Failed to bind to Custom Tabs service")
        }

        activity.application.registerActivityLifecycleCallbacks(activityLifecycleCallbacks)
    }

    private fun getEngagementSignalsCallback(
        browserSwitchCallbacks: BrowserSwitchCallbacks?
    ): EngagementSignalsCallback =
        object : EngagementSignalsCallback {
            override fun onVerticalScrollEvent(isDirectionUp: Boolean, extras: Bundle) {
                Log.d("asdf", "onVerticalScrollEvent (isDirectionUp=$isDirectionUp)")
            }

            override fun onGreatestScrollPercentageIncreased(scrollPercentage: Int, extras: Bundle) {
                Log.d("asdf", "scroll percentage: $scrollPercentage%")
            }

            override fun onSessionEnded(didUserInteract: Boolean, extras: Bundle) {
                browserSwitchCallbacks?.onFinished?.invoke()
            }
        }


    private val activityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
        override fun onActivityStarted(activity: Activity) {}
        override fun onActivityResumed(activity: Activity) {}
        override fun onActivityPaused(activity: Activity) {}
        override fun onActivityStopped(activity: Activity) {}
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) {
            if (activity == targetActivity) {
                cleanUp(activity)
            }
        }
    }

    private fun cleanUp(activity: Activity) {
        customTabsServiceConnection?.let {
            activity.unbindService(it)
            customTabsServiceConnection = null
        }
        targetActivity = null
        Log.d("asdf", "Service unbound and cleaned up")
    }
}
